enhancements/upgrade-bundle.md

```diff
@@ -0,0 +1,258 @@
+
+Currently, the run command has different subcommands: 
+
+ * `run local`
+ * `run packagemanifests`
+ * `run bundle` - currently being implemented, design document is covered by [run bundle enhancement](https://github.com/operator-framework/enhancements/blob/master/enhancements/run-bundle.md)
+
+For doing an `bundle-upgrade`, there could be two scenarios: 
+
+* Case 1 - A previous version of the operator is already running on the cluster, and the index image contains this operator version. Now, we we want to upgrade the in cluster running to a new version that we specify. 
+
+* Case 2 - No previous version of an operator is running on the cluster, so we provide the new bundle image which is accessible and pullable within the cluster, which contains our desired version of the operator for the upgrade. 
+
+In case 1, we will have to figure out which catalog source the existing operator was installed in and somehow dynamically inject the new bundle image into the index image for OLM to detect this change and upgrade to the new version we provided. 
+
+In this proposal, we will design bundle upgrade for case 2. For this case, we also assume that an index image already exists that has the from version of the operator bundle. So we don't create a new index image. We use the existing index image (which is mandatory to provide) for the operator and provide the new bundle image for the upgrade. For this to happen, we need to specify starting CSV.
+Generally, when there is no starting CSV specified, OLM always detects and installs the latest version available for an operator in a particular  channel. For example, in production case, if an operator is running and there's a subscription for it, and there comes a fix available in a stable channel, OLM will detect and upgrade by default. 
+
+Let's consider a few scenarios: 
+
+How do we verfiy whether an operator has been upgraded successfully?
+To verify whether an operator upgrade to a desired version has been successful, the ClusterServiceVersion (CSV) CustomResource is a useful resource for providing such context to the end user. A given subscription has its status field and corresponding InstallPlan. If a ClusterServiceVersion (CSV) is created for that InstallPlan, then the CSV Status indicates that the operator has been successfully installed or upgraded. 
+  
+In case of subscription failure, the end user needs to be aware of this. If a subscription failed, there might not be an InstallPlan that's created for the new operator bundle. However, through the SubscriptionStatus, we can obtain the CurrentCSV and compare it with the desired operator version we wanted to upgrade to, and verify that they are not the same, and hence conclude there's been a failure. 
+We then log a message that "run bundle-upgrade was not successful", clean u the registry pod and exit.
```

Comment:
It might be useful to keep the registry pod there to maybe determine the graph that was being used. I think that flag to keep this on failure might be helpful?

Reply:
Yes, great point! I don't see any reason why the registry pod alone needs to be cleaned up. We could use a --skip-cleanup flag like scorecard to offer flexibility to operator developers for clean up options. 
Updated it.